### PR TITLE
App-shared port without DNS server is not necessarily an issue

### PR DESCRIPTION
**Pull request by milan-zededa**
Not having DNS servers configured (while link is OK and IP address is assigned), is not necessarily a failure for an app-shared interface. Some applications might not require DNS for external hostnames and may only use internal DNS servers to resolve the names of other apps running on the same device. However, we should still issue a warning about this potential issue.

### CODE REVIEW

**Comment by milan-zededa** [comment]
@OhmSpectator It seems that we have caught some panic in the PSI unit test:
```
=== FAIL: agentlog TestPsiEveIntegratedStartTwice (10.08s)
    agentlog_test.go:624: could not start the integrated psi collector API: could not start the server in 10 seconds

=== FAIL: agentlog TestPsiEveIntegratedStopUnstarted (unknown)
	github.com/lf-edge/eve/pkg/pillar/agentlog	coverage: 24.8% of statements
panic: test timed out after 10m0s
running tests:
	TestPsiEveIntegratedStopUnstarted (9m43s)
```
https://github.com/lf-edge/eve/actions/runs/11123159400/job/30906006512?pr=4315

**Comment by OhmSpectator** [comment]
This code fails:
```
	go agentlog.ListenDebug(logObj, "stackdump", "memdump")
	started := false
	// Wait for ListenDebug to start http service (try to get 200 response)
	// Limit the number of retries to avoid infinite loop, try for 10 seconds
	for i := 0; i < 100; i++ {
		resp, err := http.Get("http://127.0.0.1:6543")
		if err == nil && resp.StatusCode == 200 {
			started = true
			break
		}
		time.Sleep(100 * time.Millisecond)
	}
	if !started {
		return nil, fmt.Errorf("could not start the server in 10 seconds")
	
```

May it be somehow related to the changes in the PR?

@christoph-zededa, do we have other unit tests for the Debug Server? It fails to start for some reason...

**Comment by OhmSpectator** [comment]
@milan-zededa, could you reproduce it locally?
Usually for that, I do:
```
cd pkg/pillar
make enter-docker-dev
... it enters container here
(container) cd pillar/agentlog
(container) go test -v
```

**Comment by christoph-zededa** [comment]
> @christoph-zededa, do we have other unit tests for the Debug Server? It fails to start for some reason...

We have only https://github.com/lf-edge/eve/blob/master/pkg/pillar/agentlog/http-debug_test.go#L66

Also I would not call this a unit test ...

**Comment by OhmSpectator** [comment]
> > @christoph-zededa, do we have other unit tests for the Debug Server? It fails to start for some reason...
> 
> We have only https://github.com/lf-edge/eve/blob/master/pkg/pillar/agentlog/http-debug_test.go#L66
> 
> Also I would not call this a unit test ...

Ooph... Can we automate all the steps mentioned in the comments?.. Ok, @milan-zededa, this one is also to be checked if the failure is consistent... I'll try locally as well.

**Comment by christoph-zededa** [comment]
Also I see:
```
2024-10-01T10:20:24.2670978Z ✖  agentlog (10m0.131s)
```
So the test times out ..., too.

**Comment by OhmSpectator** [comment]
> Also I see:
> 
> ```
> 2024-10-01T10:20:24.2670978Z ✖  agentlog (10m0.131s)
> ```
> 
> So the test times out ..., too.

Only in this PR, or the master as well?

**Comment by christoph-zededa** [comment]
> Only in this PR, or the master as well?

I doubt it is in master otherwise we would have seen failed tests.

**Comment by OhmSpectator** [comment]
> > Only in this PR, or the master as well?
> 
> I doubt it is in master otherwise we would have seen failed tests.

Then, this PR somehow affected our debug server started by Pillar...

**Comment by christoph-zededa** [comment]
> Then, this PR somehow affected our debug server started by Pillar...

Or the test is flaky ;-)

**Comment by OhmSpectator** [comment]
I tried it locally and it ran successfully...

**Review by OhmSpectator — COMMENTED** [review]
I restarted the tests here, and they also work fine...

> ✓  agentlog (24.775s)

Hm... Interesting; what would be a reason for this long server start?... @christoph-zededa, could we add a basic HttpDebug test that we can run in CI?

**Comment by christoph-zededa** [comment]
> I restarted the tests here, and they also work fine...
> 
> > ✓  agentlog (24.775s)
> 
> Hm... Interesting; what would be a reason for this long server start?... @christoph-zededa, could we add a basic HttpDebug test that we can run in CI?

seems it was in this mutex here: https://github.com/lf-edge/eve/actions/runs/11123159400/job/30906006512?pr=4315#step:3:583 (agentlog_test.go:408 )

I am not sure if another test helps, your test is already testing it as well and testing that the golang http server works is already covered by their (=golang standard lib) tests enough.

**Comment by OhmSpectator** [comment]
Also, can the lock still be locked if a signal interrupts the previous instance of the server running in a test? I'll create a ticket for that. I hate "easy go testing".

**Comment by christoph-zededa** [comment]
> Also, can the lock still be locked if a signal interrupts the previous instance of the server running in a test? 

Like f.e. SIGTERM? The mutex is only valid during runtime.

**Comment by OhmSpectator** [comment]
> > Also, can the lock still be locked if a signal interrupts the previous instance of the server running in a test?
> 
> Like f.e. SIGTERM? The mutex is only valid during runtime.

Hm... Then I have no idea how we could deadlock there... Wanna take a look? =)

**Comment by OhmSpectator** [comment]
Yeah, indeed, it was the case. The server has not started in one test, forgot to unlock the mutex, and failed the next test in deadlock. I'll fix the deadlock soon.
I still have a question: why were 10 seconds insufficient for the server to start? Should we increase it time, @christoph-zededa? Or can it be the regular problem described by @milan-zededa before of the runner that does not give control to the VM?

**Comment by christoph-zededa** [comment]
> I still have a question: why were 10 seconds insufficient for the server to start? 

Would be good to have a profile or more log messages to see what is going on.

**Comment by OhmSpectator** [comment]
> Would be good to have a profile or more log messages to see what is going on.

Weren't you able to reproduce it locally with the HttpDebug test you mentioned before?

**Comment by OhmSpectator** [comment]
@milan-zededa Sorry for discussing all that time a failure in the HttpDebug test)) I'll also take a look at the code of the PR itself))

File: pkg/pillar/types/conntest.go [hunk 1]
```diff
@@ -40,18 +51,25 @@ func (trPtr *TestResults) HasError() bool {
 	return trPtr.LastFailed.After(trPtr.LastSucceeded)
 }
 
+// HasWarning returns true if test succeeded but there is a warning reported.
+func (trPtr *TestResults) HasWarning() bool {
+	return !trPtr.HasError() && trPtr.LastWarning != ""
```
Comment (OhmSpectator) [inline pkg/pillar/types/conntest.go:56]: Shouldn't the same logic be used here as used in the `HasError()`? Something like 
```
return trPtr.LastWarned.After(trPtr.LastSucceeded) && ...
```

**Review by OhmSpectator — APPROVED** [review]
Running the Eden tests.

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/pillar/types/conntest.go (hunk 1 above)
Comment (milan-zededa) [reply pkg/pillar/types/conntest.go:56]: I don't think that is necessary. Warning is used when connectivity test is successful but there is some potential issue with the config that the user should be aware of. So we continue using the `LastSucceeded` timestamp in this case.

**Comment by milan-zededa** [comment]
Looks like tests already run and succeeded: https://github.com/lf-edge/eve/actions/runs/11134071507
No idea why they are not linked from this PR and instead it is shown as Skipped.

**Comment by OhmSpectator** [comment]
@milan-zededa, I think this one is also to be backported to 13.4-stable?

**Comment by milan-zededa** [comment]
> @milan-zededa, this one as well.. I guess we still need it in 13.4-stable.

@OhmSpectator Here is the 13.4 backport: https://github.com/lf-edge/eve/pull/4345

**Pull request merged**